DOC Add dropdowns to User Guide section 3.2, "Tuning the hyper-parameters of an estimator" (#27631)
scikit-learn/scikit-learn: doc/modules/grid_search.rst

.. topic:: Examples:

    - See :ref:`sphx_glr_auto_examples_model_selection_plot_nested_cross_validation_iris.py`
      for an example of Grid Search within a cross validation loop on the iris
      dataset. This is the best practice for evaluating the performance of a
      model with grid search.

    - See :ref:`sphx_glr_auto_examples_model_selection_plot_grid_search_text_feature_extraction.py` for an example
      of Grid Search coupling parameters from a text documents feature
      extractor (n-gram count vectorizer and TF-IDF transformer) with a
      classifier (here a linear SVM trained with SGD with either elastic
      net or L2 penalty) using a :class:`~sklearn.pipeline.Pipeline` instance.

    |details-start|
    Advanced examples
    |details-split|

    - See :ref:`sphx_glr_auto_examples_model_selection_plot_multi_metric_evaluation.py`
      for an example of :class:`GridSearchCV` being used to evaluate multiple
      metrics simultaneously.

    - See

>       :ref:`sphx_glr_auto_examples_model_selection_plot_grid_search_refit_callable.py`

Author: those 2 examples seem to showcase very similar functionality

  is set to 'exhaust'. This means that `min_resources` is automatically set
  such that the last iteration can use as many resources as possible, within
  the `max_resources` limit::

      >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
      ...                          factor=2, min_resources='exhaust').fit(X, y)
      >>> sh.n_resources_
      [250, 500, 1000]

  `min_resources` was here automatically set to 250, which results in the last
  iteration using all the resources. The exact value that is used depends on
  the number of candidate parameter, on `max_resources` and on `factor`.

  For :class:`HalvingRandomSearchCV`, exhausting the resources can be done in 2
  ways:

  - by setting `min_resources='exhaust'`, just like for
    :class:`HalvingGridSearchCV`;
  - by setting `n_candidates='exhaust'`.

  Both options are mutually exclusive: using `min_resources='exhaust'` requires
  knowing the number of candidates, and symmetrically `n_candidates='exhaust'`
  requires knowing `min_resources`.

  In general, exhausting the total number of resources leads to a better final
  candidate parameter, and is slightly more time-intensive.

.. _aggressive_elimination:

> .. dropdown:: Aggressive elimination of candidates

Reviewer: This section is being called [HalvingRandomSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.HalvingRandomSearchCV.html) and [HalvingGridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.HalvingGridSearchCV.html) docstrings. We rather keep the section for traceability.

```suggestion
Aggressive elimination of candidates
------------------------------------

Using the ``aggressive_elimination`` parameter, you can force the search
process to end up with less than ``factor`` candidates at the last
iteration.

.. dropdown:: Code example of aggressive elimination
```

  then just have to pick the best one. When the number of available resources is
  small with respect to the number of candidates, the last iteration may have to
  evaluate more than ``factor`` candidates::

      >>> from sklearn.datasets import make_classification
      >>> from sklearn.svm import SVC
      >>> from sklearn.experimental import enable_halving_search_cv  # noqa
      >>> from sklearn.model_selection import HalvingGridSearchCV
      >>> import pandas as pd
      >>> param_grid = {'kernel': ('linear', 'rbf'),
      ...               'C': [1, 10, 100]}
      >>> base_estimator = SVC(gamma='scale')
      >>> X, y = make_classification(n_samples=1000)
      >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
      ...                          factor=2, max_resources=40,
      ...                          aggressive_elimination=False).fit(X, y)
      >>> sh.n_resources_
      [20, 40]
      >>> sh.n_candidates_
      [6, 3]

  Since we cannot use more than ``max_resources=40`` resources, the process
  has to stop at the second iteration which evaluates more than ``factor=2``
  candidates.

>   Using the ``aggressive_elimination`` parameter, you can force the search
>   process to end up with less than ``factor`` candidates at the last
>   iteration. To do this, the process will eliminate as many candidates as
>   necessary using ``min_resources`` resources::

Reviewer: ```suggestion
  When using ``aggressive_elimination``, the process will eliminate as many
  candidates as necessary using ``min_resources`` resources::
```

      >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
      ...                            factor=2,
      ...                            max_resources=40,
      ...                            aggressive_elimination=True,
      ...                            ).fit(X, y)
      >>> sh.n_resources_
      [20, 20, 40]
      >>> sh.n_candidates_
      [6, 3, 2]

  Notice that we end with 2 candidates at the last iteration since we have
  eliminated enough candidates during the first iterations, using ``n_resources =
  min_resources = 20``.

.. _successive_halving_cv_results:

> .. dropdown:: Analyzing results with the `cv_results_` attribute

Reviewer: This section is being called [HalvingRandomSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.HalvingRandomSearchCV.html) and [HalvingGridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.HalvingGridSearchCV.html) docstrings. We rather keep the section for traceability.

```suggestion
Analyzing results with the `cv_results_` attribute
--------------------------------------------------

The ``cv_results_`` attribute contains useful information for analyzing the
results of a search. It can be converted to a pandas dataframe with ``df =
pd.DataFrame(est.cv_results_)``. The ``cv_results_`` attribute of
:class:`HalvingGridSearchCV` and :class:`HalvingRandomSearchCV` is similar
to that of :class:`GridSearchCV` and :class:`RandomizedSearchCV`, with
additional information related to the successive halving process.
```

>   The ``cv_results_`` attribute contains useful information for analyzing the
>   results of a search. It can be converted to a pandas dataframe with ``df =
>   pd.DataFrame(est.cv_results_)``. The ``cv_results_`` attribute of
>   :class:`HalvingGridSearchCV` and :class:`HalvingRandomSearchCV` is similar
>   to that of :class:`GridSearchCV` and :class:`RandomizedSearchCV`, with
>   additional information related to the successive halving process.
> 
>   Here is an example with some of the columns of a (truncated) dataframe:

Reviewer: ```suggestion
.. dropdown:: Example of a (truncated) output dataframe:
```

Robustness to failure
---------------------

> Some parameter settings may result in a failure to ``fit`` one or more folds of

Author: This is not related to dropdowns but that part appears to be outdated as `error_score=np.nan` is now the default; LMK if it should be handled in another issue & pr instead

Reviewer: Nice catch! I think it's ok to address the issue here.
